File: LinkedLists/Python/Inserciones.py
#Creamos la clase nodo en donde damos la estructura de lo que contiene un nodo con dos constructores disponibles
class Node:
    def __init__(self): #Constructor vacio
        self.data = None #Dato que almacenara el nodo
        self.next = None #Nodo siguiente
    
    def __init__(self, data): #Constructor que recibe el valor del nodo
        self.data = data
        self.next = None

#Creamos la clase lista la cual contendra la cabeza de la lista y un atributo para llevar el control del tamano de la lista
class LinkedList:
    def __init__(self):
        self.head = Node(None)
        self.size = 0

#Funcion para insertar nodo al inicio de la lista
#Esta recibira la lista, el nodo cabeza, y el nuevo dato a insertar
#Complejidad Temporal: O(1)
#Complejidad Espacial Auxiliar: O(1)
def push(list, head, data):
    #Creamos el nuevo nodo
    new_node = Node(data)
    #Este nuevo nodo tiene que ir de primero, por lo que ahora este debe de apuntar al primer nodo de la lista, es decir, al siguiente de la cabeza
    new_node.next = head.next
    #Ahora, la cabeza debe de apuntar al primer nodo, es decir, al nuevo nodo creado
    head.next = new_node


#Funcion para insertar al final de la lista
#Esta recibira la lista , la referencia al nodo encabezador, y el nuevo dato a insertar
#Complejidad Temporal: O(n)
#Complejidad Espacial Auxiliar: O(1)
def append(List, head, data):
    #Creamos el nuevo nodo
    new_node = Node(data)
    #Si la lista es nula
    if head == None : return
    #Si la lista esta vacia, se insertara el nuevo nodo como primero, es decir, el siguiente de la cabeza
    if head.next == None:
        head.next = new_node
    else: #Sino, iteramos hasta encontrar el ultimo nodo, es decir, cuando el siguiente del nodo en el que nos encontramos sea Nulo
        #Creamos una copia de la cabeza, para poder iterar sobre la lista, ya que sino, perderemos la referencia original de la cabeza
        last = head
        #Iteramos hasta que el siguiente del nodo en el que estamos sea nulo, es decir, hasta cuando estemos en el ultimo nodo
        while last.next is not None:
            last = last.next
        #Cuando acaba el ciclo anterior, sabemos que estamos parados sobre el ultimo nodo
        #Como queremos insertar un nuevo nodo al final, el siguiente del nodo en el que estamos debe de ser el nuevo nodo creado
        last.next = new_node
    #Finalmente incrementamos el valor de la longitud de la lista
    List.size += 1

#Funcion para insertar al final de la lista recursivamente
#Complejidad Temporal: O(n)
#Complejidad Espacial: O(1)
#Esta funcion debe de recibir el puntero de la lista, la referencia al nodo encabezador, y la referencia del nodo a insertar
def _append(List, head, new_node):
    #Case Base
    #Si la cabeza es nula retornamos, o si el siguiente de cabeza es nulo, insertamos en nuevo nodo como primero, es decir, el siguiente de cabeza
    if head is None : return
    if head.next is None:
        head.next = new_node
        #Incrementamos la longitud de la lista
        List.size += 1
        return
    else: #Sino, llamamos recursivamente la funcion, hasta que se cumpla el caso base, en donde sabremos que llegamos al ultimo nodo, y ahi insertaremos
        #Creamos una copia de la cabeza para poder recorrer la lista sin perder la referencia de la cabeza
        last = head
        #Hacemos el llamado recursivo, en donde pasaremos como cabeza, el siguiente de la cabeza actual
        _append(List, last.next, new_node)

#Funcion para insertar en la lista en una posicion dada
#Complejidad Temporal: O(n)
#Complejidad Espacial: O(1)
#Esta funcion recibira la lista, el nodo cabeza, el dato a insertar, y la posicion donde se quiere insertar
def insert(List, head, data, pos):
    #Creamos el nuevo nodo
    new_node = Node(data)
    #Si la posicion es mayor a la longitud de la lista o menor o igual a cero retornamos error
    if pos > List.size or pos <= 0 :
        print("La posicion no es valida")
        return
    #Si la posicion es uno, insertamos al inicio de la lista
    #Es decir, el nuevo nodo sera el siguiente de la cabeza
    if pos == 1 :
        new_node.next = head.next
        head.next = new_node
    else:
        #Sino, iteraremos por toda la lista hasta llegar al nodo que este en la posicion pasada y alli insertar
        #Creamos un contador el cual ayudara a saber la posicion del nodo en el que estemos parados
        cont = 1
        #Crearemos una copia de el siguiente de la cabeza para asi iterar, y estar en el nodo que corresponde al contador
        curr = head.next
        #Necesitamos siempre tener la referencia al nodo previo del que estamos parados, en este caso;
        prev = head
        #Iteramos hasta que curr que es nuestro nodo actual llegue al final de la lista:
        while curr is not None:
            #Preguntamos si la posicion es igual al contador
            if cont == pos:
                #Si entramos, significa que nuestro nodo actual, esta en la posicion pasada
                #Aqui es donde debemos de insertar el nuevo nodo
                #Para esto, necesitamos que el previo al nodo actual, apunte a el nuevo nodo
                prev.next = new_node
                #Luego, el siguiente de nuestro nuevo nodo, debe ser el nodo actual
                new_node.next = curr
                #Incrementamos la longitud de la lista
                List.size += 1
                return
            else:
                #Si lo anterior no se cumple, debemos de incrementar el contador y a su vez, pasar los dos nodos auxiliares a sus siguientes
                cont += 1
                prev = curr
                curr = curr.next

#Ahora crearemos una algoritmo para insertar elementos en una lista con una posicion dada pero de forma recursiva
#Para realizar esto necesitaremos dos funciones:
#Una que reciba la lista, la cabeza, el nodo a insertar, y la posicion. En esta haremos un par de verificaciones, y luego crearemos 3 variables, un contador para llevar la cuenta de en que nodo estamos, un nodo actual que sera el siguiente de la lista, y un nodo previo al actual que nos servira como referencia a la hora de insertar el nuevo nodo
#Complejidad Temporal: O(n)
#Complejidad Espacial: O(n)
#Primera funcion:
def _insert(List, head, new_node, pos):
    #Si la posicion es mayor a la longitud de la lista o menor o igual a cero retornamos error
    if pos > List.size or pos <= 0 :
        print("La posicion no es valida")
        return
    #Si la posicion es uno, insertamos al inicio de la lista
    #Es decir, el nuevo nodo sera el siguiente de la cabeza
    if pos == 1 :
        new_node.next = head.next
        head.next = new_node
        return
    #Ahora , si lo anterior no se cumple, crearemos un contador que nos servira para saber en que posicion de la lista estamos.
    cont  = 1 
    #Ahora crearemos un nodo que nos servira como nodo actual el cual estara ubicado inicialmente en el primer elemento de la lista. Con este recorreremos la lista hasta que este en la posicion que necesitamos
    curr = head.next
    #Ahora necesitamos un nodo que nos sirva como previo al nodo actual creado anteriormente
    prev = head
    #Con estos datos creados , podemos llamar la funcion que los recibira y hara la busqueda del nodo en la posicion recursivamente e insertara alli
    __insert(List, head, new_node, prev, curr, cont, pos)
#Segunda Funcion
def __insert(List, head, new_node, prev, curr, cont, pos):
    #Caso base:
    #Si el contador es igual a la posicion, debemos de insertar alli el nodo con ayuda de prev y curr
    if cont  == pos:
        prev.next = new_node
        new_node.next = curr
        List.size += 1
        return
    else:
        #Sino , debemos de incrementar el contador, y avanzar los dos nodos auxiliares a su siguiente
        cont += 1
        prev = curr
        curr = curr.next
        #Llamamos recursivamente la funcion con los nuevos datos
        __insert(List, head, new_node, prev, curr, cont, pos)

#-----------------------------------------------------------

File: LinkedLists/#1.Singly-Linked-List/C++/Inserciones.cpp
#include<bits/stdc++.h>
using namespace std;

//Creamos la clase nodo en donde damos la estructura de lo que contiene un nodo con dos constructores disponibles
class Node{
    public:
        int data; //Dato que contiene el nodo
        Node *next; //Puntero al siguiente nodo
        Node(){this->data = 0 ; this->next = NULL;}
        Node(int data){this->data = data ; this->next = NULL;}
};

//Creamos la clase lista la cual contendra la cabeza de la lista y un atributo adicional para llevar un control del tamano de la lista
class LinkedList{
    public:
        int size; //Atributo que se ira actualizando con el tamano de la lista 
        Node *head; //Creamos un puntero hacia el nodo cabeza de nuestra lista
        LinkedList(){this->size = 0 ; this->head = new Node();} //Constructor
};


//Funcion para insertar al inicio de la lista.
//Esta funcion se puede implementar recibiendo como tal la referencia al nodo cabeza
//Complejidad Temporal : O(1)
//Complejidad Espacial: O(1)
void push(LinkedList *list, Node *&head_ref, int data ){
    //Creamos el nuevo nodo a insertar
    Node *new_node = new Node(data);
    //Decimos que el siguiente del nuevo nodo , es el que es ahora el siguiente de la cabeza
    new_node->next = head_ref->next;
    //Actualizamos que el siguiente de la cabeza es el nuevo nodo
    head_ref->next = new_node;
}

//Funcion para insertar al final de la lista.
//Complejidad Temporal : O(n)
//Complejidad Espacial : O(1)
void append(LinkedList *list, Node *&head, int data){
    //Creamos el nuevo nodo
    Node *new_node = new Node(data);
    //Si la lista es nula
    if (head == NULL){return;}
    //Si la lista esta vacia, insertamos el nuevo nodo como el siguiente de la cabeza, es decir, sera el primero de la lista
    if(head->next == NULL){
        head->next = new_node;
    }else{ //Sino iteramos sobre la lista para buscar el ultimo nodo
        //Creamos una variable temporal que apunte a cabeza, para asi poder desplazarnos por la lista y no modificar la referencia de la cabeza
        Node *last = head;
        //Iteramos hasta saber que last es el ultimo nodo, es decir, hasta que su siguiente sea nulo
        while(last->next != NULL){
            last = last->next;
        }
        //Cuando acaba el ciclo, ya sabemos que last esta apuntando al ultimo nodo de la lista
        //Como queremos apuntar al final, lo que toca hacer es que este ultimo nodo, apunte como siguiente al nuevo nodo
        last->next = new_node;
    }
    //Incrementamos la longitud de la lista
    list->size++;
}

//Funcion para insertar al final recursivamente
//Complejidad Temporal: O(n)
//Complejidad Espacial: O(n)
//Esta funcion debe de recibir el puntero de la lista, la referencia al nodo encabezador, y la referencia del nodo a insertar
void _append(LinkedList *list, Node *&head, Node*& new_node){
    //Caso Base
    //Si la cabeza es nula retornamos, o la lista esta vacia, pegamos el nuevo nodo de primero en la lista, es decir, sera el siguiente de cabeza y retornamos
    if(head == NULL) return;
    if(head->next == NULL){
        head->next = new_node;
        list->size++;
        return;
    }else{ //Sino, recorreremos recursivamente la lista, hasta que se cumpla el caso base, si llega ahi es porque encontro el ultimo nodo ya que su siguiente es nulo, y alli pega el nuevo nodo
        //Creamos copia de la cabeza, para no perder su referencia
        Node *last = head;
        //Hacemos llamada recursiva en donde ya la cabeza, sera el siguiente de la cabeza actual
        _append(list, last->next , new_node);
    }
}

//Funcion para insertar un elemento en la lista en una posicion dada
//Complejidad Temporal: O(n)
//Complejidad Espacial: O(1)
//Esta funcion recibira el puntero de la lista, la referencia al nodo cabeza, el elemento a insertar, y la posicion en donde se quiere insertar
void insert(LinkedList *list, Node *&head, int data, int pos){
    //Creamos el nuevo nodo
    Node *new_node = new Node(data);
    if(pos > list->size || pos <= 0 ){
        cout<<"Se esta accediendo a una posicion erronea"<<endl;
        return;
    }
    //Si la posicion es uno, insertamos al inicio de la lista
    //Es decir, el nuevo nodo sera el siguiente de cabeza
    if(pos == 1 ){
        new_node->next = head->next;
        head->next = new_node;
    }else{ //Sino, iteraremos hasta encontrar el nodo que corresponda con la posicion pasada
        //Crearemos contador para saber la posicion del nodo en el que estamos
        int cont = 1 ;
        //Creamos una copia del siguiente de la cabeza para estar en la primera posicion de la lista
        Node *curr = head->next;
        //Creamos un nodo previo a curr, el cual nos servira para tener una referencia al nodo previo
        Node *prev = head;
        //Iteramos hasta que el nodo curr, no llegue al final de la lista, es decir , hasta cuando sea Nulo
        while(curr != NULL){
            //Preguntamos si el contador esta en la posicion pasada
            if (cont == pos){
                //El siguiente de previo sera el nuevo nodo creado
                prev->next = new_node;
                //El siguiente del nuevo nodo sera curr
                new_node->next = curr;
                //Incrementamos la longitud de la lista
                list->size++;
                return;
            }else{
                //Si no estamos en la posicion, incrementamos contador y aumentamos cada uno de los nodos auxiliares a su siguiente
                cont++;
                prev = curr;
                curr = curr->next;
            }
        }

    }
}

//Funcion para insertar en la lista en una posicion dada recursivamente
//Para implementar esto necesitamos dos funciones, la primera recibe la lista, la cabeza, el nodo a insertar y la posicion a insertar
//Alli hace verificaciones y si todo esta bien , crea tres variables necesarias
//Un contador para llevar la cuenta de que nodo estamos, un nodo actual y un nodo previo
//De alli llama la segunda funcion la cual es la que hace el proceso recursivo para insertar el nuevo nodo en la posicion pasada.
//Complejidad Temporal: O(n)
//Complejidad Espacial: O(n)
//Declaremos primero las funciones:
void __insert(LinkedList *, Node *&, Node *&, int);
void _insert(LinkedList *, Node *&, Node *&, Node *&, Node *&, int, int);
//Primera Funcion:
void __insert(LinkedList *list, Node *&head, Node *&new_node, int pos){
    //Verificamos que la posicion pasada sea valida
    //Es decir, no sea ni mayor ni menos a la longitud de la lista
    if(pos > list->size || pos <= 0 ){
        cout<<"La posicion ingresada es erronea"<<endl;
        return;
    }
    //Si la posicion es uno, insertamos al inicio de la lista
    if(pos == 1){
        new_node->next = head->next;
        head->next = new_node;
        return;
    }
    //Creamos variable contador que se ubicara en el primero elemento de la lista, e ira incrementando hasta llegar a la posicion dada
    int cont = 1;
    //Crearmos la variable nodo que nos permitira iterar sobre la lista
    //Esta debe de decir el primer elemento de la lista, ya que este debe de concordar con la posicion dada con el contador siempre
    Node *curr = head->next;
    //Necesitamos una variable nodo que siempre sea el previo de la variable actual (curr)
    Node *prev = head;
    //Aca, ya podemos llamar a la funcion _insert la cual recibira todos estos todas para asi insertar el nuevo nodo de forma recursiva
    _insert(list, head, new_node, prev, curr, cont, pos);
}
//Segunda Funcion:
//Esta funcion recibira el puntero a la lista, la referencia del nodo cabeza, la referencia del nodo que se va a insertar, un nodo llamado previo el cual lo recibiremos inicialmente apuntando a la cabeza, otro nodo llamado curr el cual lo recibiremos apuntando al siguiente de cabeza,el contador  y la posicion
void _insert(LinkedList *list, Node *&head, Node *&new_node, Node *&prev , Node*&curr,int cont,  int pos){
    //Caso Base:
    //Si la posicion es igual al contador, insertamos y retornamos
    if(cont == pos){
        //El siguiente del previo es el nuevo nodo
        prev->next = new_node;
        //El siguiente del nuevo nodo, es el nodo actual
        new_node->next = curr;
        //Incrementamos el valor de la longitud de la lista
        list->size++;
        return;
    }else{
        //Incrementamos contador
        cont++;
        //Pasamos prev y curr a sus siguientes
        prev = curr;
        curr = curr->next;
        //Pasamos recursivamente los datos
        _insert(list, head, new_node, prev, curr,cont, pos);
    }
}
//---------------------------------------------------------------


